hatch/app.py

```diff
@@ -1,39 +1,54 @@
 import aiofiles
-from fastapi import Depends, FastAPI, Header, HTTPException, Security
+from fastapi import Depends, FastAPI, HTTPException, Security
 from fastapi.responses import FileResponse
 from fastapi.security.api_key import APIKeyHeader
-from pydantic import Required
+from pydantic import ValidationError
+from starlette.requests import Request
 
-from hatch import config, models
+from hatch import config, models, schema
+from hatch.signing import AuthToken, set_default_key
 
 
+app = FastAPI()
+# set key to use in signing
+set_default_key(config.BACKEND_TOKEN, config.BACKEND)
 api_key_header = APIKeyHeader(name="Authorization")
 
 
-def validate(auth_token: str = Security(api_key_header)):
-    if auth_token != config.BACKEND_TOKEN:
+def validate(request: Request, auth_token: str = Security(api_key_header)):
+    try:
+        token = AuthToken.verify(auth_token)
+    except ValidationError:
         raise HTTPException(403, "Unauthorised")
 
+    if not str(request.url).startswith(token.url):
+        raise HTTPException(403, "Unauthorised")
 
-app = FastAPI(dependencies=[Depends(validate)])
+    return token
 
 
-@app.get(
-    "/workspace/{workspace}",
-    response_model=models.FilesIndex,
-)
-def workspace_index(workspace: str):
-    """Return an index of the files on disk in this workspace."""
+def validate_workspace(workspace):
+    """Validate a workspace exists on disk."""
     path = config.WORKSPACES / workspace
-
     if not path.exists():
         raise HTTPException(404, f"Workspace {workspace} not found")
 
-    return models.FilesIndex.from_dir(path, f"/workspace/{workspace}/")
+    return path
 
 
-@app.get("/workspace/{workspace}/{name:path}")
-async def workspace_file(workspace: str, name: str):
+@app.get("/workspace/{workspace}/current/", response_model=schema.IndexSchema)
+def workspace_index(
+    workspace: str, request: Request, token: AuthToken = Depends(validate)
```

Comment:
Does the `token: AuthToken = Depends(validate)` kwarg mean `token` is of type `AuthToken` and has a default of the `Depends` call or that it calls `Depends` here?

Reply:
So, this is a form of Dependency Injection, using type annotation defaults.

AIUI, fastapi looks at the function arg annotations, checks the defaults. If any are `Depends()`, it evaluates them in the request context and sets the arg's value to the result.

It works a bit like pytest's fixture injection, including fixture dependencies, but a lot less magically.

I actually quite like it, I think.

FTR, I moved from using `validate()` at the app level bakck to validating at the individual API level because I want the `token` as a function variable, mainly so I can access `token.user` and `token.scope` in the view body.